File: services/user-service/models/playlist.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Playlist represents a user's video playlist
type Playlist struct {
	ID          uint           `json:"id" gorm:"primaryKey"`
	UserID      uint           `json:"user_id" gorm:"not null;index"`
	Name        string         `json:"name" gorm:"size:100;not null"`
	Description string         `json:"description" gorm:"size:500"`
	IsPublic    bool           `json:"is_public" gorm:"default:false"`
	VideoCount  int            `json:"video_count" gorm:"default:0"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
	
	// Relationships
	Videos []PlaylistVideo `json:"videos,omitempty" gorm:"foreignKey:PlaylistID;constraint:OnDelete:CASCADE"`
}

// PlaylistVideo represents the relationship between playlists and videos
type PlaylistVideo struct {
	ID         uint           `json:"id" gorm:"primaryKey"`
	PlaylistID uint           `json:"playlist_id" gorm:"not null;index"`
	VideoID    uint           `json:"video_id" gorm:"not null"`
	Position   int            `json:"position" gorm:"not null;default:0"`
	AddedAt    time.Time      `json:"added_at"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`
	
	// Relationships
	Playlist Playlist `json:"playlist,omitempty" gorm:"foreignKey:PlaylistID"`
}

// PlaylistWithVideos represents a playlist with video metadata
type PlaylistWithVideos struct {
	ID          uint                    `json:"id"`
	UserID      uint                    `json:"user_id"`
	Name        string                  `json:"name"`
	Description string                  `json:"description"`
	IsPublic    bool                    `json:"is_public"`
	VideoCount  int                     `json:"video_count"`
	CreatedAt   time.Time               `json:"created_at"`
	UpdatedAt   time.Time               `json:"updated_at"`
	Videos      []PlaylistVideoWithMeta `json:"videos"`
}

// PlaylistVideoWithMeta represents a playlist video with video metadata
type PlaylistVideoWithMeta struct {
	ID         uint      `json:"id"`
	PlaylistID uint      `json:"playlist_id"`
	VideoID    uint      `json:"video_id"`
	Position   int       `json:"position"`
	AddedAt    time.Time `json:"added_at"`
	
	// Video metadata (fetched from Video Service)
	VideoTitle       string `json:"video_title,omitempty"`
	VideoDescription string `json:"video_description,omitempty"`
	VideoDuration    int    `json:"video_duration,omitempty"`
	VideoThumbnail   string `json:"video_thumbnail,omitempty"`
	VideoCategoryID  uint   `json:"video_category_id,omitempty"`
	VideoViewCount   int    `json:"video_view_count,omitempty"`
}

// TableName overrides the table name for Playlist
func (Playlist) TableName() string {
	return "playlists"
}

// TableName overrides the table name for PlaylistVideo
func (PlaylistVideo) TableName() string {
	return "playlist_videos"
}

// Request/Response types

// CreatePlaylistRequest represents the request body for creating a playlist
type CreatePlaylistRequest struct {
	Name        string `json:"name" validate:"required,min=1,max=100"`
	Description string `json:"description" validate:"max=500"`
	IsPublic    bool   `json:"is_public"`
}

// UpdatePlaylistRequest represents the request body for updating a playlist
type UpdatePlaylistRequest struct {
	Name        string `json:"name" validate:"min=1,max=100"`
	Description string `json:"description" validate:"max=500"`
	IsPublic    *bool  `json:"is_public"`
}

// AddVideoToPlaylistRequest represents the request body for adding a video to playlist
type AddVideoToPlaylistRequest struct {
	VideoID  uint `json:"video_id" validate:"required"`
	Position *int `json:"position"` // Optional, will append to end if not specified
}

// ReorderPlaylistVideosRequest represents the request body for reordering playlist videos
type ReorderPlaylistVideosRequest struct {
	VideoOrders []VideoOrder `json:"video_orders" validate:"required,dive"`
}

// VideoOrder represents the new order for a video in playlist
type VideoOrder struct {
	VideoID  uint `json:"video_id" validate:"required"`
	Position int  `json:"position" validate:"min=0"`
}

// PlaylistListResponse represents the response for listing playlists
type PlaylistListResponse struct {
	Playlists []Playlist `json:"playlists"`
	Total     int        `json:"total"`
	Page      int        `json:"page"`
	Limit     int        `json:"limit"`
	Pages     int        `json:"pages"`
}

// PlaylistVideosResponse represents the response for playlist videos
type PlaylistVideosResponse struct {
	Playlist PlaylistWithVideos `json:"playlist"`
	Total    int                `json:"total"`
	Page     int                `json:"page"`
	Limit    int                `json:"limit"`
	Pages    int                `json:"pages"`
}

File: mobile/StreamLiteMobile/src/types/playlist.ts
// Playlist-related TypeScript type definitions

export interface Playlist {
  id: number;
  user_id: number;
  name: string;
  description: string;
  is_public: boolean;
  video_count: number;
  created_at: string;
  updated_at: string;
  first_video_thumbnail?: string;
}

export interface PlaylistVideo {
  id: number;
  playlist_id: number;
  video_id: number;
  position: number;
  added_at: string;
}

export interface PlaylistVideoWithMeta {
  id: number;
  playlist_id: number;
  video_id: number;
  position: number;
  added_at: string;
  
  // Video metadata from Video Service
  video_title?: string;
  video_description?: string;
  video_duration?: number;
  video_thumbnail?: string;
  video_category_id?: number;
  video_view_count?: number;
}

export interface PlaylistWithVideos {
  id: number;
  user_id: number;
  name: string;
  description: string;
  is_public: boolean;
  video_count: number;
  created_at: string;
  updated_at: string;
  videos: PlaylistVideoWithMeta[];
}

// Request/Response types
export interface CreatePlaylistRequest {
  name: string;
  description?: string;
  is_public?: boolean;
}

export interface UpdatePlaylistRequest {
  name?: string;
  description?: string;
  is_public?: boolean;
}

export interface AddVideoToPlaylistRequest {
  video_id: number;
  position?: number;
}

export interface VideoOrder {
  video_id: number;
  position: number;
}

export interface ReorderPlaylistVideosRequest {
  video_orders: VideoOrder[];
}

// API Response types
export interface PlaylistListResponse {
  playlists: Playlist[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface PlaylistVideosResponse {
  playlist: PlaylistWithVideos;
  total: number;
  page: number;
  limit: number;
  pages: number;
}

// API Query parameters
export interface PlaylistListParams {
  page?: number;
  limit?: number;
}

export interface PlaylistVideosParams {
  page?: number;
  limit?: number;
}

// Error types
export interface PlaylistError {
  message: string;
  code?: string;
  field?: string;
}

export interface PlaylistApiResponse<T = any> {
  data?: T;
  error?: PlaylistError;
  success: boolean;
}

// UI State types
export interface PlaylistFormData {
  name: string;
  description: string;
  isPublic: boolean;
}

export interface PlaylistModalState {
  visible: boolean;
  mode: 'create' | 'edit' | 'view';
  playlist?: Playlist;
}

export interface AddToPlaylistModalState {
  visible: boolean;
  videoId?: number;
  videoTitle?: string;
}

// Playlist player state
export interface PlaylistPlayerState {
  currentPlaylist?: PlaylistWithVideos;
  currentVideoIndex: number;
  isPlaying: boolean;
  shuffle: boolean;
  repeat: 'none' | 'one' | 'all';
  autoPlay: boolean;
}

// Playlist management actions
export interface PlaylistAction {
  type: 'CREATE' | 'UPDATE' | 'DELETE' | 'ADD_VIDEO' | 'REMOVE_VIDEO' | 'REORDER';
  playlistId?: number;
  videoId?: number;
  data?: any;
}

// Playlist filter and sort options
export interface PlaylistFilterOptions {
  showPublicOnly?: boolean;
  showPrivateOnly?: boolean;
  sortBy?: 'name' | 'created_at' | 'updated_at' | 'video_count';
  sortOrder?: 'asc' | 'desc';
}

// Playlist sharing types
export interface PlaylistShareData {
  playlistId: number;
  playlistName: string;
  shareUrl: string;
  isPublic: boolean;
}

// Playlist statistics
export interface PlaylistStats {
  totalPlaylists: number;
  totalVideos: number;
  totalDuration: number; // in seconds
  averageVideosPerPlaylist: number;
  mostPopularPlaylist?: Playlist;
}
